8239493: Add classfile generation to jextract

JornVernee opened the pull request:
Hi,

This patch adds class file generation to jextract using ASM, which in some places replaces the generation done by the current source code generator. In particularly, it replaces the generation of all constant fields with static getters, that use dynamic constants as backing storage rather than fields. This allows for more sharing between constants, and also reduces the pressure on clinit code size a lot, enabling extraction of much bigger native header files. The new scheme generates an extra constant helper class which holds all these constants, and the source generator generates the high-level wrapper functions. The scheme can also be extended to generate multiple constant helper classes, should we run out of constant pool indices (in rare cases).

This patch reworks the interface of JavaSourceBuilder, since extra parameters are needed in order to do the new constant generation. For generating methods that need constants, it calls into a new ConstantHelper class, which returns a DirectMethodHandleDesc describing the generated constant getter. A call to this getter is then emitted.

This patch also merges the HandleSourceFactory and StaticWrapperSourceFactory into OutputFactory, since only the latter was ever being used, but the split in code made it a lot harder to reason about the correctness of for instance changing C identifiers into java-safe ones. Because of this, the diff for OuputFactory looks a bit messy, so maybe it's easier to look at the resulting file as a whole.

The tests are changed to look for the now generated static getters, instead of fields (`static final` fields don't work well with dynamic constants, since they are eagerly initialized).

As a small bonus, this patch adds the missing ABI-agnostic C_LONGDOUBLE constant (this was causing some test failures).

Thanks,
Jorn
<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [JDK-8239493](https://bugs.openjdk.java.net/browse/JDK-8239493): Add classfile generation to jextract


### Reviewers
 * Maurizio Cimadamore ([mcimadamore](@mcimadamore) - Committer)
 * Athijegannathan Sundararajan ([sundar](@sundararajana) - Committer) ⚠️ Review applies to 517a9c7212fcb050ca404e8191bdfc6ac2cac59e

### Download
`$ git fetch https://git.openjdk.java.net/panama-foreign pull/102/head:pull/102`
`$ git checkout pull/102`

mcimadamore commented on src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/ConstantHelper.java:
```diff
@@ -0,0 +1,476 @@
+            }
+        } else if (type.isPrimitive()) {
+            if (type == int.class || type == byte.class || type == short.class || type == char.class) {
+                return emitGetter(name, type, mv -> emitConInt(mv, ((Long) value).intValue()));
```
I think having `emitIntConstantGetter`, `emitFloatConstantGetter` could be clearer. I noted that emitGetter is always used either through the condy case (which is fine) or through one of those constants getters here - I think *hiding* the lambda might make the code more readable here (and then you can keep existing general `emiteGetter` as an helper that is not to be used directly).

JornVernee commented:
I've addressed review comments, and also added another test that tests some of the new constant support for strings and memory addresses (among others).

JornVernee commented on src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/resources/RuntimeHelper.java.template:
```diff
@@ -52,6 +53,22 @@ public class RuntimeHelper {
         return lookup(LIBRARIES, name).orElse(null);
     }
 
+    public static final MemoryAddress makeCString(String value) {
```
After offline discussion, I've removed this and am now calling Cstring::toCString

JornVernee commented on src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/JavaSourceBuilder.java:
```diff
@@ -24,350 +24,175 @@
-        sb.append(";\n");
-        decrAlign();
+    private String getFunction(String javaName, FunctionDescriptor fDesc) {
+        return getterCall(constantHelper.addFunctionDesc(javaName, fDesc));
```
I've renamed these to `addMethodHandleGetter` and `methodHandleGetCallString` for more clarity, as well as moving the latter to the end of the file to avoid confusion.

JornVernee commented:
/integrate
